### PR TITLE
[cpu-offload] Enable NUMA by default

### PR DESCRIPTION
**Pull request by mlazos**
Two fixes for CPU activation offloading on GB200 (Blackwell)
  hardware:

  run_train.sh: Add --numa-binding node to torchrun — Pins each
  worker process to the NUMA node containing its assigned GPU. On
  GB200 NVLink-C2C, this guarantees ~350 GB/s D2H/H2D bandwidth for
  CPU offload. Without it, the OS scheduler can migrate threads
  cross-NUMA, degrading bandwidth to ~120 GB/s.

Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #3523
* __->__ #3489

### CODE REVIEW

**Review by tianyu-l — CHANGES_REQUESTED** [review]
Apologize for blocking, as
- I couldn't tell what this code does
- it touches core

Maybe you meant a draft?

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 1]
```diff
@@ -424,10 +424,18 @@ def apply_cpu_offload_pass(
...
+        last_use_arg = last_consumer
+        last_use_val = last_consumer.meta.get("val")
+        if not isinstance(last_use_val, torch.Tensor):
+            last_use_arg = None
+
         with gm.graph.inserting_after(offload_node):
             wait_offload_node = gm.graph.call_function(
                 torch.ops.ao.wait_tensor.default,
-                args=(offload_node, node, last_consumer),
+                args=(offload_node, node, last_use_arg),
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]:
```
@torch.library.impl("ao::wait_tensor", "CompositeExplicitAutograd")
def _ao_wait_tensor(
    tensor: torch.Tensor,
    keepalive: torch.Tensor | None = None,
    last_use_of_storage: torch.Tensor | None = None,
) -> torch.Tensor:
    completion_event, device = _pop_wait(tensor)
    current_stream = torch.accelerator.current_stream(device)

    current_stream.wait_event(completion_event)
    if keepalive is not None:
        storage = keepalive.untyped_storage()
        if storage.size() > 0:
            storage.resize_(0)
    return tensor

```

last_use_of_storage is not used as as computation, and it's only used as an edge to enforce topo order. 

In theory, it shouldn't matter what's passed in here. 

What's actually failing? is it custom op schema checker or something?

File: run_train.sh [hunk 2]
```diff
@@ -42,6 +42,7 @@ else
     PYTORCH_ALLOC_CONF="expandable_segments:True" \
     TORCHFT_LIGHTHOUSE=${TORCHFT_LIGHTHOUSE} \
     torchrun --nproc_per_node=${NGPU} --rdzv_backend c10d --rdzv_endpoint="localhost:0" \
+    --numa-binding node \
```
Comment (SherlockNoMad) [inline run_train.sh]: what the default value of torchrun numa-binding? from torchrun doc, it's unclear. 

Also if the default value is not "node", then this change is meaningful, and should also be mention in a few more places, e.g. cpu_offload code and claude.md, so that agent always pickup, and remember to set numa-binding.

File: torchtitan/experiments/graph_trainer/cpu_offload.py (hunk 1 above)
Comment (mlazos) [reply torchtitan/experiments/graph_trainer/cpu_offload.py]: yeah custom op schema checker, I asked claude this same question haha

**Comment by mlazos** [comment]
> Apologize for blocking, as
> 
> * I couldn't tell what this code does
> * it touches core
> 
> Maybe you meant a draft?

Added PR desc, sorry about that.

File: run_train.sh (hunk 2 above)
Comment (mlazos) [reply run_train.sh]: yeah default is None, then you can end up having cross-node allocations with that setting.

Comment (tianyu-l) [reply run_train.sh]: - do we know why the default is not node?
- does it work on all hardware, or only cuda?
- maybe add comments nearby
- also, we probably should change internal mast launcher script as well

File: torchtitan/experiments/graph_trainer/cpu_offload.py (hunk 1 above)
Comment (SherlockNoMad) [reply torchtitan/experiments/graph_trainer/cpu_offload.py]: if it's schema, we should fix schema

File: run_train.sh (hunk 2 above)
Comment (SherlockNoMad) [reply run_train.sh]: +1

Comment (mlazos) [reply run_train.sh]: Chatted with sherlock offline, I'm going to programmatically do this instead, the numa option will fail on non-cuda hardware, so when we initialize training I'll setup the numa config using the pytorch core API for doing this (what torchrun calls) when we are on NV hw.

File: torchtitan/experiments/graph_trainer/cpu_offload.py (hunk 1 above)
Comment (mlazos) [reply torchtitan/experiments/graph_trainer/cpu_offload.py]: Moved into next PR, did as we discussed with grabbing the getitem child

**Review by SherlockNoMad — APPROVED** [review]
lgtm

**Comment by tianyu-l** [comment]
how's this different to the other PR which I left comment https://github.com/pytorch/torchtitan/pull/3519#discussion_r3353640873

**Comment by mlazos** [comment]
> how's this different to the other PR which I left comment [#3519 (comment)](https://github.com/pytorch/torchtitan/pull/3519#discussion_r3353640873)

Sorry, closed the other one, claude messed up some ghstack metadata. This is the actual PR. Well there isn't downside to enabling this in the eager trainer as well, but I can do that. Wdyt?

**Comment by tianyu-l** [comment]
> Well there isn't downside to enabling this in the eager trainer as well, but I can do that. Wdyt?

Yeah I would wait for the upside for core trainer to put it there.

**Comment by mlazos** [comment]
> > Well there isn't downside to enabling this in the eager trainer as well, but I can do that. Wdyt?
> 
> Yeah I would wait for the upside for core trainer to put it there.

Updated, I put it in graph trainer.

**Review by tianyu-l — APPROVED** [review]
